servicetalk-http-utils/src/main/java/io/servicetalk/http/utils/JavaNetSoTimeoutHttpConnectionFilter.java

```diff
@@ -167,24 +173,101 @@ public Single<StreamingHttpResponse> request(final StreamingHttpRequest request)
-                                            .onErrorMap(TimeoutException.class, t -> newStacklessSocketTimeoutException(
-                                                    "Read timed out after " + timeout.toMillis() +
-                                                            "ms waiting for response meta-data")
-                                                    .initCause(t))))
-                            .map(response -> response.transformMessageBody(p -> p.timeout(timeout, timeoutExecutor)
-                                    .onErrorMap(TimeoutException.class, t -> newStacklessSocketTimeoutException(
-                                            "Read timed out after " + timeout.toMillis() +
-                                                    "ms waiting for the next response payload body chunk")
-                                            .initCause(t))))
+                            .<StreamingHttpResponse>liftSync(subscriber ->
+                                    new RequestTimeoutSubscriber(subscriber,
+                                            SourceAdapters.fromSource(requestProcessor), timeout, timeoutExecutor))
                             .shareContextOnSubscribe();
                 });
             }
         };
     }
 
+    // package private for testing purposes
+    static final class RequestTimeoutSubscriber implements Subscriber<StreamingHttpResponse> {
+
+        private static final AtomicIntegerFieldUpdater<RequestTimeoutSubscriber> onceUpdater =
+                AtomicIntegerFieldUpdater.newUpdater(RequestTimeoutSubscriber.class, "once");
+
+        private final DelayedCancellable requestCancellable = new DelayedCancellable();
+        private final DelayedCancellable timeoutCancellable = new DelayedCancellable();
+        private final Subscriber<? super StreamingHttpResponse> delegate;
+
+        private final Completable requestComplete;
+        private final Duration timeout;
+        private final Executor timeoutExecutor;
+        @SuppressWarnings("unused")
+        private volatile int once;
+
+        RequestTimeoutSubscriber(Subscriber<? super StreamingHttpResponse> delegate, Completable requestComplete,
+                                        Duration timeout, Executor timeoutExecutor) {
+            this.delegate = delegate;
+            this.requestComplete = requestComplete;
+            this.timeout = timeout;
+            this.timeoutExecutor = timeoutExecutor;
+        }
+
+        @Override
+        public void onSubscribe(Cancellable cancellable) {
+            delegate.onSubscribe(() -> {
+                // We don't need to condition cancellation here on the `once()` call because the `DelayedCancellable`
+                // will already enforce idempotency of the cancel call, and it's fine if we're racing cancels with the
+                // results since cleanup is gated on the `once()` call.
```

Comment:
It doesn't matter how many times `cancel()` will be invoked for the timer and request, spec rule 3.7 says that the original `Subscription` must become NOP. So, it's responsibility of the original source.

Reply:
Not necessarily to be removed completely, but may need to be updated if you remove `DelayedCancellable` for `timeoutCancellable` based on the other comments